app/src/main/java/org/projectbuendia/client/ui/chart/PatientChartController.java

```diff
@@ -269,21 +269,21 @@ public void suspend() {
         }
     }
 
-    public void onXFormResult(int requestCode, int resultCode, Intent data) {
-        FormRequest request = popFormRequest(requestCode);
-        if (request == null) {
+    public void onXFormResult(final int requestCode, final int resultCode, final Intent data) {
+        final FormRequest request = popFormRequest(requestCode);
+         if (request == null) {
             LOG.e("Unknown form request code: " + requestCode);
             return;
         }
 
-        boolean shouldShowSubmissionDialog = (resultCode != Activity.RESULT_CANCELED);
-        String action = (resultCode == Activity.RESULT_CANCELED)
-            ? "form_discard_pressed" : "form_save_pressed";
-        Utils.logUserAction(action,
-            "form", request.formUuid,
-            "patient_uuid", request.patientUuid);
-        mOdkResultSender.sendOdkResultToServer(request.patientUuid, resultCode, data);
-        mUi.showFormSubmissionDialog(shouldShowSubmissionDialog);
+        final boolean isSubmissionCanceled = (resultCode == Activity.RESULT_CANCELED);
+        Utils.logUserAction(isSubmissionCanceled ? "form_discard_pressed" : "form_save_pressed",
+            "form", request.formUuid, "patient_uuid", request.patientUuid);
+        if(isSubmissionCanceled) return;
```

Comment:
I'm not sure we're talking about the same thing :)

I meant that now that you're checking whether submission was cancelled here (`if(isSubmissionCanceled) return;`), you can remove the identical check in OdkResultSender#sendOdkResultToServer, on the line `if(isActivityCanceled(resultCode, data)) return false;`.

Does that make sense?